token_filter.module

```diff
@@ -64,9 +64,123 @@ function _token_filter_filter_tokens($text, $filter, $format, $langcode, $cache,
     $data['user'] = $user;
   }
 
-  return token_replace($text, $data, $options);
+  // The rest of this function comes from token_replace(), but we insert our own
+  // additional filtering of the tokens to be replaced.
+  $text_tokens = token_scan($text);
+  if (empty($text_tokens)) {
+    return $text;
+  }
+
+  $replace = $filter->settings['replace'];
```

Comment:
This throws a `Notice: Undefined index: replace in _token_filter_filter_tokens() (line 74 of XXX/modules/token_filter/token_filter.module).`

To reproduce:
1. In the Configure filter modal, leave the Tokens textfield empty (no tokens forbidden)
2. "Update" and "Save configuration"
3. Create a node with some tokens in the body
4. Save. Clear the cache

See the notice.

 ```suggestion
  $replace = !empty($filter->settings['replace']) ? $filter->settings['replace'] : array() ;
```

Reply:
Actually, my suggestion is completely wrong (too early in the morning!). I'm not able to reproduce this again. I'm not sure what happened

Reply:
@argiepiano @bugfolder 
I think I've reproduced something related to this.  Follow the steps above. Go to admin/content/node  Clear "Page and else" cache (think clearing all caches would also do it but narrowed down to this).

Got warnings:
> Warning: Undefined array key "evaluate" in _token_filter_filter_tokens() (line 74 of /app/docroot/modules/token_filter/token_filter.module).

> Warning: Undefined array key "tokens" in _token_filter_filter_tokens() (line 85 of /app/docroot/modules/token_filter/token_filter.module).

> Deprecated function: preg_quote(): Passing null to parameter #1 ($str) of type string is deprecated in preg_quote() (line 85 of /app/docroot/modules/token_filter/token_filter.module).

If I revert the patch then this error goes away (spent ages trying to figure this out in case a module I'm porting was causing it 🤦 )

Reply:
@yorkshire-pudding you may be working with an old or wrong version of token_filter. Line 74 in the dev version is a comment. Be sure to apply this patch to the dev version. The patch will not apply correctly to the latest release.

Reply:
Sorry, my bad. I was looking at the dev versino before applying the patch. Line 74 does exist.

However, I'm not able to reproduce. I've followed the steps.

Reply:
I've tried to reproduce on vanilla site but no joy. Must be something else interacting with it as that site has a number of modules I maintain or am porting or attempted to port so I'll have to work out which it is.

Reply:
Bingo! I was able to reproduce. 

The issue happens because, when you first enable the filter after a clean installation (that is, a virgin site that has never had Token Filter installed before - this is related to another minor bug I'll report in the issue), then enable he filter WITHOUT editing and saving the "Configure" modal =  then this is what you get in the config
```json
        "filter_tokens": {
            "status": 1,
            "weight": 0,
            "module": "token_filter",
            "settings": []
        }
```

As you can see, settings is empty, which throws the warnings. I think a solution might be to initialize with default settings. Normally this should look like this:

```json
        "filter_tokens": {
            "status": 1,
            "weight": 0,
            "settings": {
                "replace": "0",
                "tokens": ""
            },
            "module": "token_filter"
        }
```

Reply:
Confirmed that config did look like that and saving the config fixes it